### PR TITLE
Convert exponential numbers to string ecma-262/5.1/#sec-9.8.1

### LINKED ISSUE #155: jsfuck issue...
```
>>> js2py.eval_js('(+(+!+[]+[+!+[]]+(!![]+[])[!+[]+!+[]+!+[]]+[!+[]+!+[]]+[+[]])+[])[+!+[]]')
u'1'
>>>
```

should translate to `.` instead it translated to `1`

### PR DESCRIPTION
**Pull request by pro-src**
Fixes #155

### CODE REVIEW

**Comment by pro-src** [comment]
I'm not concerned with the codeclimate issue but the travis-ci build can probably be fixed by checking for zero prior to calling `log10`. We might always want to convert to float before that condition as well.

**Comment by pro-src** [comment]
@VeNoMouS you want to give this a try?

File: js2py/base.py [hunk 1]
```diff
@@ -608,6 +609,16 @@ def to_string(self):
             elif self.is_infinity():
                 sign = '-' if self.value < 0 else ''
                 return Js(sign + 'Infinity')
+            elif not (-7 < floor(log10(abs(self.value))) < 21):
+                # Remove the sign from negative zero
+                x = 0 if self.value == 0 else self.value
```
Comment (pro-src) [inline js2py/base.py]: The zero check can be removed from here. It's leftovers from `Number.prototype.toExponential`.

File: js2py/base.py [hunk 2]
```diff
@@ -608,6 +609,16 @@ def to_string(self):
...
+                n = format(x, '-.%de' % 16).split('e')
+                # Remove trailing zeros
+                s = n[0].rstrip('.0') or '0'
```
Comment (pro-src) [inline js2py/base.py]: This can be just `s = n[0].rstrip('0')

**Comment by VeNoMouS** [comment]
@pro-src  breaks on py2 and py3 ..

```
Python 3.5.3 (default, Sep 27 2018, 17:25:39)
[GCC 6.3.0 20170516] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import js2py
>>> js2py.eval_js('(+(+!+[]+[+!+[]]+(!![]+[])[!+[]+!+[]+!+[]]+[!+[]+!+[]]+[+[]])+[])[+!+[]]')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/tmp/Js2Py/js2py/evaljs.py", line 115, in eval_js
    return e.eval(js)
  File "/tmp/Js2Py/js2py/evaljs.py", line 188, in eval
    self.execute(code, use_compilation_plan=use_compilation_plan)
  File "/tmp/Js2Py/js2py/evaljs.py", line 183, in execute
    exec (compiled, self._context)
  File "<EvalJS snippet>", line 2, in <module>
  File "/tmp/Js2Py/js2py/base.py", line 967, in __call__
    return self.call(self.GlobalObject, args)
  File "/tmp/Js2Py/js2py/base.py", line 1481, in call
    return Js(self.code(*args))
  File "/tmp/Js2Py/js2py/host/jseval.py", line 45, in Eval
    executor(py_code)
  File "/tmp/Js2Py/js2py/host/jseval.py", line 51, in executor
    exec (code, globals())
  File "<string>", line 2, in <module>
  File "/tmp/Js2Py/js2py/base.py", line 794, in __add__
    b = other.to_primitive()
  File "/tmp/Js2Py/js2py/base.py", line 548, in to_primitive
    return self.default_value(hint)
  File "/tmp/Js2Py/js2py/base.py", line 452, in default_value
    cand = method.call(self)
  File "/tmp/Js2Py/js2py/base.py", line 1481, in call
    return Js(self.code(*args))
  File "/tmp/Js2Py/js2py/prototypes/jsarray.py", line 26, in toString
    return func.call(this, ())
  File "/tmp/Js2Py/js2py/base.py", line 1481, in call
    return Js(self.code(*args))
  File "/tmp/Js2Py/js2py/prototypes/jsarray.py", line 82, in join
    s = elem.to_string().value
  File "/tmp/Js2Py/js2py/base.py", line 612, in to_string
    elif not (-7 < floor(log10(abs(self.value))) < 21):
ValueError: math domain error
```

**Comment by pro-src** [comment]
@VeNoMouS Ya, this is trash :(

I wrote a test generator: https://gist.github.com/pro-src/c9d871e0138a9e041b2c2162d240ad7c
The results are not great at all but give me some time and I'll work something out. This proving the be a bit hard for me to tackle but I'm not quite deterred yet. So we'll see.

**Comment by PiotrDabkowski** [comment]
I guess the math domain error is due to 0. Let me know if the problem turns out to be too hard, will help/implement myself.

**Comment by pro-src** [comment]
@PiotrDabkowski have you had a chance to review the updated PR?

**Comment by pro-src** [comment]
Ping.

**Pull request merged**